sympy/functions/combinatorial/factorials.py

```diff
@@ -1095,7 +1095,7 @@ def _eval_rewrite_as_factorial(self, n, k, **kwargs):
 
     def _eval_rewrite_as_gamma(self, n, k, **kwargs):
         from sympy import gamma
-        if n.is_nonnegative and n.is_integer or n.is_integer is False:
+        if not all(x.is_negative and x.is_integer for x in (n, k, n - k)):
```

Comment:
I am uncomfortable with this. This would allow a rewrite for a vanilla symbol since vanilla.is_negative is None, the `all` would fail, the `not all` would pass and the rewrite would return and there would be nothing to stop someone from substituting the vanilla(s) with negatives and then the formula wouldn't be right. Of course even with the right assumptions there is nothing to stop the user from doing that, but then it is their fault for doing the substitution.

💡  It occurs to me that we could have a safe option for subs using the `assumptions` function: if there is any discrepancy between non-None values in the old and new expressions then an error could be raised.

Reply:
Ahhh okay.
in that case I'll just make sure that n, k and n - k are non negative integers or gamma would fail or shall we let it fail, they want it rewritten as gamma right, and gamma doesn't support -ve ints.
>💡 It occurs to me that we could have a safe option for subs using the assumptions function: if there is any discrepancy between non-None values in the old and new expressions then an error could be raised.

Yeah we could do that, but that'd be a new pr right ? we'd have to test the change extensively.

Reply:
> they want it rewritten as gamma right

IMO the rewrite should only be returned if it is valid; if it can't be rewritten in some form because of conditions then we should not return it. I don't know if there are rewrite routines that raise an error if the conditions are wrong or return a Piecewise giving smthng like `Piecewise((what you want, conditions), (nan, True))`. The only difference with this and the invalid gamma is that they will see that there are conditions rather than learn of them post-subs.

Reply:
> we'd have to test the change extensively.

the default would be `safe=False` so there would only be minimal tests needed. But yes...another PR.

Reply:
> > they want it rewritten as gamma right
> 
> IMO the rewrite should only be returned if it is valid; if it can't be rewritten in some form because of conditions then we should not return it. 
> ...
> The only difference with this and the invalid gamma is that they will see that there are conditions rather than learn of them post-subs.

This is reasonable. we can make this happen. I'll get some sleep and start workin on it later today.

Reply:
> we could have a safe option for subs

Ahh yes. We already have `check_assumptions` but it will raise if checking an expression against a non-symbol:
```
        if not isinstance(against, Symbol):
            raise TypeError('against should be of type Symbol')
```
So that TypeError can be removed and replaced with assumptions = _assumptions(against)` where `from sympy.core.assumptions import assumptions as _assumptions`.